homeassistant/components/light/tplink.py

```diff
@@ -119,30 +125,51 @@ def rgb_color(self):
 
     @property
     def is_on(self):
-        """True if device is on."""
+        """Return True if device is on."""
         return self._state
 
     def update(self):
         """Update the TP-Link Bulb's state."""
         from pyHS100 import SmartDeviceException
         try:
+            if self._supported_features == 0:
+                self.get_features()
             self._state = (
                 self.smartbulb.state == self.smartbulb.BULB_STATE_ON)
-            self._brightness = brightness_from_percentage(
-                self.smartbulb.brightness)
-            if self.smartbulb.is_color:
+            if self._name is None:
+                self._name = self.smartbulb.alias
+            if self.smartbulb.is_dimmable:
+                self._brightness = brightness_from_percentage(
+                    self.smartbulb.brightness)
+            if self.smartbulb.is_variable_color_temp:
                 if (self.smartbulb.color_temp is not None and
                         self.smartbulb.color_temp != 0):
                     self._color_temp = kelvin_to_mired(
                         self.smartbulb.color_temp)
+            if self.smartbulb.is_color:
                 self._rgb = hsv_to_rgb(self.smartbulb.hsv)
+            if self.smartbulb.has_emeter:
+                self._emeter_params[ATTR_CURRENT_CONSUMPTION] \
+                    = "%.1f W" % self.smartbulb.current_consumption()
+                daily_statistics = self.smartbulb.get_emeter_daily()
+                monthly_statistics = self.smartbulb.get_emeter_monthly()
+                self._emeter_params[ATTR_DAILY_CONSUMPTION] \
+                    = "%.2f kW" % daily_statistics[int(time.strftime("%d"))]
+                self._emeter_params[ATTR_MONTHLY_CONSUMPTION] \
+                    = "%.2f kW" % monthly_statistics[int(time.strftime("%m"))]
         except (SmartDeviceException, OSError) as ex:
-            _LOGGER.warning('Could not read state for %s: %s', self.name, ex)
+            _LOGGER.warning('Could not read state for %s: %s', self._name, ex)
 
     @property
     def supported_features(self):
         """Flag supported features."""
-        supported_features = SUPPORT_TPLINK
+        return self._supported_features
+
+    def get_features(self):
+        """Determine all supported features in one go."""
+        if self.smartbulb.is_dimmable:
+            self._supported_features += SUPPORT_BRIGHTNESS
+        if self.smartbulb.is_variable_color_temp:
+            self._supported_features += SUPPORT_COLOR_TEMP
         if self.smartbulb.is_color:
-            supported_features += SUPPORT_RGB_COLOR
-        return supported_features
+            self._supported_features += SUPPORT_RGB_COLOR
```

Comment:
Just wanted to point out here that this will induce in total of 3 requests on the device, as `self.sysinfo` is not cached by pyhs100. I don't think it matters much, but worth noticing I think.

Reply:
This is 3 calls in get_features, which should be called once in the lifetime, should be fine. Should be a lot better than them being called in supported_features as they were before.

Reply:
Thanks, looks good to me! :+1: